include/engine/routing_algorithms/routing_base.hpp

```diff
@@ -727,20 +732,60 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
         }
     }
 
+    bool NeedsLoopForward(const PhantomNode &source_phantom,
+                          const PhantomNode &target_phantom) const
+    {
+        return source_phantom.forward_node_id == target_phantom.forward_node_id &&
+               source_phantom.GetForwardWeightPlusOffset() >
+                   target_phantom.GetForwardWeightPlusOffset();
+    }
+
+    bool NeedsLoopBackwards(const PhantomNode &source_phantom,
+                            const PhantomNode &target_phantom) const
+    {
+        return source_phantom.reverse_node_id == target_phantom.reverse_node_id &&
+               source_phantom.GetReverseWeightPlusOffset() >
+                   target_phantom.GetReverseWeightPlusOffset();
+    }
+
+    double GetPathDistance(const std::vector<NodeID> &packed_path,
+                           const PhantomNode &source_phantom,
+                           const PhantomNode &target_phantom) const
+    {
+        std::vector<PathData> unpacked_path;
+        PhantomNodes nodes;
+        nodes.source_phantom = source_phantom;
+        nodes.target_phantom = target_phantom;
+        UnpackPath(packed_path.begin(), packed_path.end(), nodes, unpacked_path);
```

Comment:
Or even better, we do not need the vector from above at all, it's wasted allocations.

What about accumulating the distance on the fly by using `function_output_iterator`?

- http://www.boost.org/doc/libs/1_60_0/libs/iterator/doc/function_output_iterator.html
- see rtd/include/io/stream.h#L123